add test cases from test project

OceanOak opened the pull request:
**Files that were added successfully:**
testering.res
fib.res
effect_time.res
main_bookreader.res
main_clicker.res
main_counter_navigation.res
main_counter.res
main_entry.res
main_field.res
main_form.res
main_random_color.res
main_random.res
counterParts.res
counter.res
main_embeddedCounters.res
main_embeddedCounterParts.res
main_todo_optimized.res
main_todo_optimizedarray.res
main_todo_optimizedmap.res
main_todo.res

**changes done to solve the following error in all files that contains "todo" in their name** 

error: tagger This has type: {.."keyCode": int} => option<'a>
Somewhere wanted: Tea_json.Decoder.t<'b, 'c>

```
let onEnter = (~key="", msg) => {
  let tagger = ev =>
    switch ev["keyCode"] {
    | 13 => Some(msg)
    | _ => None
    }

  on("keydown", ~key, tagger)
}
```
**to solve this error:**
**1. added tea_html.res a different "on" function** 

`let on' = (typ, ~key="", cb) => on'(typ, key, cb)

`**tea_html.resi**

`let on': (string, ~key: string=?, Web.Node.event => option<'a>,) => Vdom.property<'a>
`
**2.Added to vdom.res** 

`let on' = (name, key:string, cb) => Events(name, key, cb)`

**add to type property<'msg> different  Event' ** 

```
type property<'msg> =
  | NoProp
  | RawProp(string, string)
  | Attribute(string, string, string)
  | Data(string, string)
  | Event(string, eventHandler<'msg>, ref<option<eventCache<'msg>>>)
  | Event'(string, string, Web.Node.event => option<'msg>)
  | Style(list<(string, string)>)
```

and because of the new Event had to add the following line of code in:

renderToHtmlString, 
patchVNodesOnElems_PropertiesApply_Add,
patchVNodesOnElems_PropertiesApply_Remove, 
patchVNodesOnElems_PropertiesApply_RemoveAdd

`| @implicit_arity Event'(_, _, _) => ()`
 
Is it ok to add it or is there a better solution?


**Other errors** 
onClick that exists in rescript-tea accepts only one argument the one used in the test files needs two arguments 
To fix this error I added:

  `let onClick' = (~key="", msg) => onCB("click", key, _ev => Some(msg))`

Same error with onDoubleClick and onBlur

```
let onDoubleClick' = (~key="", msg) => onCB("dblclick", key, _ev => Some(msg))

let onBlur' = (~key="", msg) => onCB("blur", key, _ev => Some(msg))
```

**Fixed deprecated function that were used**

**Test files that couldn't be added :**
main_clock_svg.res and main_clock.res

error in "subscriptions"
```
let main = standardProgram({
  init: init,
  update: update,
  view: view,
  subscriptions: subscriptions,
})
```
error:This call is missing an argument of type (~key: string)
This is the definition of subscriptions
```
let subscriptions = _model =>
  /* let () = Js.log ("Subscriptions", model) in */
  Time.every(Time.inMilliseconds(16.0), t => Time(t))
```

**main_gtpollute.res**
Printf.sprintf deprecated couldn't find a replacement.
Same subscriptions problem

**main_gg.res**
The value array can't be found

This array function exists in tea_json.res in the test project

```
let array = (Decoder(decoder)) => Decoder(
    value => {
      open Web.Json
      switch reify_type(value) {
      | (Array, a) =>
        let parse = v =>
          switch decoder(v) {
          | Tea_result.Ok(r) => r
          | Tea_result.Error(e) => raise(ParseFail(e))
          }
        try Tea_result.Ok(Array.map(parse, a)) catch {
        | ParseFail(e) => Tea_result.Error("Invalid array parsing: " ++ e)
        }
      | (_, _) => Tea_result.Error("Non-array value")
      }
    },
  )
```
Causes problem in reify_type can't be found
Using  Js.Types.reify_type and Js.Json.reifyType don't solve the error

**main_json.res**
same reify_type in keyvaluepaires function.

pbiggar commented on test/main_entry.res:
```diff
@@ -0,0 +1,96 @@
+/* let load = Main_counter.main (Web.Document.getElementById "content") () */
```
What's happening with this file?

OceanOak replied:
this was commented out in the bucklescript-testing repo should I remove this file ?

pbiggar replied:
What does it do?

OceanOak replied:
It starts by loading some test files using this function

example:

`let load = Main_counter.main (Web.Document.getElementById "content")`

Because the code was commented out it is still in Ocaml. Tried to uncomment it in order to covert it to rescript and see what it does exactly didn't work. 

Error when converting test/main_entry2.ml
File "", line 33, characters 1-4:
Error: Syntax error

This error is not only in this line it is everywhere.

pbiggar commented:
> 
> **to solve this error:**
> **1. added tea_html.res a different "on" function**
> 
> 
> `let on' = (typ, ~key="", cb) => on'(typ, key, cb)
> 
> 
> `**tea_html.resi**
> 
> 
> `let on': (string, ~key: string=?, Web.Node.event => option<'a>,) => Vdom.property<'a> `
> **2.Added to vdom.res**
> 
> 
> `let on' = (name, key:string, cb) => Events(name, key, cb)`
> 
> 
> **add to type property<'msg> different  Event' **
> 
> 
> ```
> type property<'msg> =
>   | NoProp
>   | RawProp(string, string)
>   | Attribute(string, string, string)
>   | Data(string, string)
>   | Event(string, eventHandler<'msg>, ref<option<eventCache<'msg>>>)
>   | Event'(string, string, Web.Node.event => option<'msg>)
>   | Style(list<(string, string)>)
> ```
> 
> 
>     
> 
>   
> 
> and because of the new Event had to add the following line of code in:
> 
> 
> renderToHtmlString,
> patchVNodesOnElems_PropertiesApply_Add,
> patchVNodesOnElems_PropertiesApply_Remove,
> patchVNodesOnElems_PropertiesApply_RemoveAdd
> 
> 
> `| @implicit_arity Event'(_, _, _) => ()`
> 
> 
> Is it ok to add it or is there a better solution?


I think the default `Vdom.on` and similar functions are supposed to have the key parameter included, but haven't. So I think the right thing is to add the key parameter to all those functions that need them.

pbiggar commented:
> **main_gtpollute.res**
> Printf.sprintf deprecated couldn't find a replacement.
> Same subscriptions problem

For this, you can use strings interpolation:

```
sprintf("%d - %s", i, str)
```

becomes

```
`${string_of_int(i)} - ${str}`
```

pbiggar commented:
> **Other errors** onClick that exists in rescript-tea accepts only one argument the one used in the test files needs two arguments To fix this error I added:
> 
> `let onClick' = (~key="", msg) => onCB("click", key, _ev => Some(msg))`
> 
> Same error with onDoubleClick and onBlur
> 
> ```
> let onDoubleClick' = (~key="", msg) => onCB("dblclick", key, _ev => Some(msg))
> 
> let onBlur' = (~key="", msg) => onCB("blur", key, _ev => Some(msg))
> ```

Similar to the `on'` situation, the extra parameter should be added to the old function instead of adding a new function.

pbiggar commented:
> **Test files that couldn't be added :** main_clock_svg.res and main_clock.res
> 
> error in "subscriptions"
> 
> ```
> let main = standardProgram({
>   init: init,
>   update: update,
>   view: view,
>   subscriptions: subscriptions,
> })
> ```
> 
> error:This call is missing an argument of type (~key: string) This is the definition of subscriptions
> 
> ```
> let subscriptions = _model =>
>   /* let () = Js.log ("Subscriptions", model) in */
>   Time.every(Time.inMilliseconds(16.0), t => Time(t))
> ```

You should also add a `key` parameter here. However, this parameter should not have a default.

pbiggar reviewed (changes requested):
This is great, nice job.

I think we can get this merged relatively quick.

OceanOak commented:
> > **to solve this error:**
> > **1. added tea_html.res a different "on" function**
> > `let on' = (typ, ~key="", cb) => on'(typ, key, cb)
> > `**tea_html.resi**
> > `let on': (string, ~key: string=?, Web.Node.event => option<'a>,) => Vdom.property<'a> `
> > **2.Added to vdom.res**
> > `let on' = (name, key:string, cb) => Events(name, key, cb)`
> > **add to type property<'msg> different  Event' **
> > ```
> > type property<'msg> =
> >   | NoProp
> >   | RawProp(string, string)
> >   | Attribute(string, string, string)
> >   | Data(string, string)
> >   | Event(string, eventHandler<'msg>, ref<option<eventCache<'msg>>>)
> >   | Event'(string, string, Web.Node.event => option<'msg>)
> >   | Style(list<(string, string)>)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and because of the new Event had to add the following line of code in:
> > renderToHtmlString,
> > patchVNodesOnElems_PropertiesApply_Add,
> > patchVNodesOnElems_PropertiesApply_Remove,
> > patchVNodesOnElems_PropertiesApply_RemoveAdd
> > `| @implicit_arity Event'(_, _, _) => ()`
> > Is it ok to add it or is there a better solution?
> 
> I think the default `Vdom.on` and similar functions are supposed to have the key parameter included, but haven't. So I think the right thing is to add the key parameter to all those functions that need them.

Vdom doesn't contain an "on" function it is my fault to name on' rather than on, I don't remember why.

**in tea.html** 

The on functions are very different, the problem is not in the key parameter.

**old**

```
  let on = (~key: string, eventName, decoder) =>
  onWithOptions(~key, eventName, defaultOptions, decoder)
```

**new**

`  let on' = (typ, ~key="", cb) => on'(typ, key, cb)`

OceanOak commented:
> > **Other errors** onClick that exists in rescript-tea accepts only one argument the one used in the test files needs two arguments To fix this error I added:
> > `let onClick' = (~key="", msg) => onCB("click", key, _ev => Some(msg))`
> > Same error with onDoubleClick and onBlur
> > ```
> > let onDoubleClick' = (~key="", msg) => onCB("dblclick", key, _ev => Some(msg))
> > 
> > let onBlur' = (~key="", msg) => onCB("blur", key, _ev => Some(msg))
> > ```
> 
> Similar to the `on'` situation, the extra parameter should be added to the old function instead of adding a new function.

onMsg accepts only one argument, adding the extra parameter will cause an error. Should I change onMsg to onCB? in this case will we keep the new functions and remove the old ones?

**old**

```
  let onClick = msg => onMsg("click", msg)
  let onDoubleClick = msg => onMsg("dblclick", msg)
  let onBlur = msg => onMsg("blur", msg)

```

**new**

```
  let onClick' = (~key="", msg) => onCB("click", key, _ev => Some(msg))
  let onDoubleClick' = (~key="", msg) => onCB("dblclick", key, _ev => Some(msg))
  let onBlur' = (~key="", msg) => onCB("blur", key, _ev => Some(msg))
```

OceanOak commented:
> > **main_gg.res** The value array can't be found
> > This array function exists in tea_json.res in the test project
> > ```
> > let array = (Decoder(decoder)) => Decoder(
> >     value => {
> >       open Web.Json
> >       switch reify_type(value) {
> >       | (Array, a) =>
> >         let parse = v =>
> >           switch decoder(v) {
> >           | Tea_result.Ok(r) => r
> >           | Tea_result.Error(e) => raise(ParseFail(e))
> >           }
> >         try Tea_result.Ok(Array.map(parse, a)) catch {
> >         | ParseFail(e) => Tea_result.Error("Invalid array parsing: " ++ e)
> >         }
> >       | (_, _) => Tea_result.Error("Non-array value")
> >       }
> >     },
> >   )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Causes problem in reify_type can't be found Using Js.Types.reify_type and Js.Json.reifyType don't solve the error
> > **main_json.res** same reify_type in keyvaluepaires function.
> 
> This confuses me because Tea_json has an `array` function: https://github.com/pbiggar/rescript-tea/blob/master/src/tea_json.resi#L24

Sorry for that, it was my fault.  This was the error in this file.
in x > 1
This has type: int
Somewhere wanted: objectTypes

let update = (model, x) =>
  switch x {
  | OnUrlChange(_map) => (model, Cmd.none)
  | AITick =>
    let width = Array.length(model.map)
    let height = Array.length(model.map[0])
    let influence = model.influence
    let units = model.units
    let tickUnit = ((x, y), ret, x) =>
      switch x {
      | Base(team, inf) =>
        /* Add in test to use up resources when building and to destroy self if not enough influence here */
        let newBase = Base(team, inf)
        if (
          x > 1 &&
            (y > 1 &&
            (influence[x - 1][y - 1] * team > 2 && false == IntPairMap.mem((x - 1, y - 1), units)))
        ) {
          list{((x - 1, y - 1), newBase), ...ret}
        } else if (
          x > 1 &&
            (y < height - 2 &&
            (influence[x - 1][y + 1] * team > 2 && false == IntPairMap.mem((x - 1, y + 1), units)))
        ) {
          list{((x - 1, y + 1), newBase), ...ret}
            
          ...
  }

OceanOak commented:
> > **Test files that couldn't be added :** main_clock_svg.res and main_clock.res
> > error in "subscriptions"
> > ```
> > let main = standardProgram({
> >   init: init,
> >   update: update,
> >   view: view,
> >   subscriptions: subscriptions,
> > })
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > error:This call is missing an argument of type (~key: string) This is the definition of subscriptions
> > ```
> > let subscriptions = _model =>
> >   /* let () = Js.log ("Subscriptions", model) in */
> >   Time.every(Time.inMilliseconds(16.0), t => Time(t))
> > ```
> 
> You should also add a `key` parameter here. However, this parameter should not have a default.

Should I add it in the subscriptions function? if so adding it will cause this error 
 This call is missing arguments of type: 'a, (~key: string)

pbiggar commented:
Great, thanks!
